linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1183,19 +1196,14 @@ private[optimizer] abstract class OptimizerCore(
                   tthenp))
             } else {
               (resolveLocalDef(tthenp), resolveLocalDef(telsep)) match {
-                case (PreTransRecordTree(thenTree, thenOrigType, thenCancelFun),
-                    PreTransRecordTree(elseTree, elseOrigType, elseCancelFun)) =>
-                  val commonType = {
-                    if (thenTree.tpe == elseTree.tpe && thenOrigType == elseOrigType)
-                      thenTree.tpe
-                    else
-                      cancelFun()
-                  }
-                  val refinedOrigType =
-                    constrainedLub(thenOrigType, elseOrigType, tree.tpe)
+                case (PreTransRecordTree(thenTree, thenStructure, thenCancelFun),
+                    PreTransRecordTree(elseTree, elseStructure, elseCancelFun)) =>
+                  if (!thenStructure.sameClassAs(elseStructure))
```

Comment:
It is true that this is much cleaner than the comparison of the record types above "in the hope" that we will not get the same two record types for different classes.